From 556c9c0a07845cc12ffb790a70b35555361a586c Mon Sep 17 00:00:00 2001
From: Razvan A Lupusoru <razvan.a.lupusoru@intel.com>
Date: Tue, 7 May 2013 14:39:56 -0700
Subject: Dalvik: Fix registerization usedef information

BZ: 104728

Registerization uses an extended MIR in order to load VRs into physical
registers. The usage of the MIR was set as defuse instead of usedef which
caused other defines of that same VR be considered dead.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I6bf7545b34f588aa05b351fe545c49967956487d
Orig-MCG-Change-Id: I31fb0e1dde0295315ec63b6910b4908fe2757e1a
Signed-off-by: Razvan A Lupusoru <razvan.a.lupusoru@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/BytecodeVisitor.cpp |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/vm/compiler/codegen/x86/BytecodeVisitor.cpp b/vm/compiler/codegen/x86/BytecodeVisitor.cpp
index ca72e50..1d38a5c 100644
--- a/vm/compiler/codegen/x86/BytecodeVisitor.cpp
+++ b/vm/compiler/codegen/x86/BytecodeVisitor.cpp
@@ -1302,7 +1302,11 @@ int getVirtualRegInfo(VirtualRegInfo* infoArray, const MIR * currentMIR) {
                 {
                     infoArray[0].regNum = currentMIR->dalvikInsn.vA;
                     infoArray[0].refCount = 2;
-                    infoArray[0].accessType = REGACCESS_DU;
+
+                    //The access type is use and then def because we use the VR when loading it into temporary
+                    //and then we alias virtual register to that temporary thus "defining" it.
+                    infoArray[0].accessType = REGACCESS_UD;
+
                     //Ok the type really depends on vB
                     RegisterClass regClass = static_cast<RegisterClass> (currentMIR->dalvikInsn.vB);
 
-- 
1.7.4.1

